ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/ContractsReader.scala

```diff
@@ -0,0 +1,111 @@
+
+/**
+  * @see [[ContractsTable]]
+  */
+private[dao] sealed abstract class ContractsReader(dispatcher: DbDispatcher) extends ContractStore {
+
+  import ContractsReader._
+
+  protected def lookupContractKeyQuery(submitter: Party, key: dao.events.Key): SimpleSql[Row]
+
+  override def lookupActiveContract(
+      submitter: Party,
+      contractId: ContractId,
+  ): Future[Option[Contract]] =
+    dispatcher.executeSql("lookup_active_contract") { implicit connection =>
+      SQL"select participant_contracts.contract_id, template_id, create_argument from #$contractsTable where contract_witness = $submitter and participant_contracts.contract_id = $contractId"
+        .as(contractRowParser.singleOpt)
+    }
+  override def lookupContractKey(
+      submitter: Party,
+      key: events.Key,
+  ): Future[Option[events.ContractId]] =
+    dispatcher.executeSql("lookup_contract_by_key") { implicit connection =>
+      lookupContractKeyQuery(submitter, key).as(contractId("contract_id").singleOpt)
+    }
+
+  override def lookupMaximumLedgerTime(ids: Set[ContractId]): Future[Option[Instant]] =
+    if (ids.isEmpty) {
+      Future.failed(emptyContractIds)
+    } else {
+      dispatcher.executeSql("lookup_maximum_ledger_time") { implicit connection =>
+        SQL"select max(create_ledger_effective_time) as max_create_ledger_effective_time, count(*) as num_contracts from participant_contracts where participant_contracts.contract_id in ($ids)"
+          .as((instant("max_create_ledger_effective_time").? ~ int("num_contracts")).single.map {
+            case result ~ retrievedContracts if retrievedContracts == ids.size => result
+            case _ => throw notFound(ids)
+          })
+      }
+    }
+
+}
+
+object ContractsReader {
+
+  private[dao] def apply(dispatcher: DbDispatcher, dbType: DbType): ContractsReader =
+    dbType match {
+      case DbType.Postgres => new Postgresql(dispatcher)
+      case DbType.H2Database => new H2Database(dispatcher)
+    }
+
+  private final class Postgresql(dispatcher: DbDispatcher) extends ContractsReader(dispatcher) {
+    override protected def lookupContractKeyQuery(
+        submitter: Party,
+        key: events.Key,
+    ): SimpleSql[Row] =
+      SQL"select participant_contracts.contract_id from #$contractsTable where $submitter =ANY(create_stakeholders) and contract_witness = $submitter and create_key_hash = ${hash(key)}"
+  }
+
+  private final class H2Database(dispatcher: DbDispatcher) extends ContractsReader(dispatcher) {
+    override protected def lookupContractKeyQuery(
+        submitter: Party,
+        key: events.Key,
+    ): SimpleSql[Row] =
+      SQL"select participant_contracts.contract_id from #$contractsTable where array_contains(create_stakeholders, $submitter) and contract_witness = $submitter and create_key_hash = ${hash(key)}"
+  }
+
+  // The contracts table _does not_ store agreement texts as they are
+  // unnecessary for interpretation and validation. The contracts returned
```

Comment:
I wasn't sure whether this was actually true and wanted to confirm that. unfortunately you won't like the result :sob:
`agreement` is a function that can be invoked during interpretation:
![image](https://user-images.githubusercontent.com/29121423/79431181-5c7c6900-7fca-11ea-9b17-017fcca0db50.png)

Reply:
Then I'm a bit concerned that this made not test fail.

Reply:
Good point. Looks like we don't have a test that uses the agreement text inside an exercise I suppose. Could you add one please?

Reply:
I was wrong in my assessment. The `agreement` expression is reified in the DAML LF template definition, and calling `agreement` in a choice body re-evaluates that expression. There is no need to store the agreement text together with the contract instance for the purpose of command interpretation.